File: mobile/lib/services/cart_service.dart
import '../services/api_config.dart';
import 'api_client.dart';

/// One line of the server's cart.
///
/// A line is a product at one sack size. The same rice at 5 kg and at 50 kg is
/// two lines, because they are two purchases at two different per-kilo prices.
class ServerCartItem {
  const ServerCartItem({
    required this.productId,
    required this.sellerId,
    required this.name,
    required this.weightKg,
    required this.quantity,
    required this.unitPrice,
    required this.lineTotal,
    required this.inStock,
    required this.availableStock,
    required this.totalKg,
    this.variety = '',
    this.image = '',
    this.priceChanged = false,
  });

  final String productId;
  final int sellerId;
  final String name;
  final String variety;
  final String image;

  /// The sack size for this line.
  final double weightKg;

  /// How many sacks - not kilograms.
  final int quantity;

  /// What one sack costs at this weight, with the tier discount applied.
  final double unitPrice;
  final double lineTotal;

  /// What these sacks come to in stock terms.
  final double totalKg;

  final bool inStock;
  final double availableStock;

  /// True when the seller changed the price after this went in the cart.
  final bool priceChanged;

  String get imageUrl => ApiConfig.mediaUrl(image);

  String get weightLabel => weightKg % 1 == 0
      ? '${weightKg.toInt()} kg'
      : '${weightKg.toStringAsFixed(1)} kg';

  /// Identifies the line for updates and removals.
  String get key => '$productId@$weightKg';

  factory ServerCartItem.fromJson(Map<String, dynamic> json) {
    return ServerCartItem(
      productId: (json['productId'] ?? '').toString(),
      sellerId: (json['sellerId'] as num?)?.toInt() ?? 0,
      name: (json['name'] ?? '').toString(),
      variety: (json['variety'] ?? '').toString(),
      image: (json['image'] ?? '').toString(),
      weightKg: (json['weightKg'] as num?)?.toDouble() ?? 1,
      quantity: (json['quantity'] as num?)?.toInt() ?? 0,
      unitPrice: (json['unitPrice'] as num?)?.toDouble() ?? 0,
      lineTotal: (json['lineTotal'] as num?)?.toDouble() ?? 0,
      totalKg: (json['totalKg'] as num?)?.toDouble() ?? 0,
      inStock: json['inStock'] != false,
      availableStock: (json['availableStock'] as num?)?.toDouble() ?? 0,
      priceChanged: json['priceChanged'] == true,
    );
  }
}

class ServerCart {
  const ServerCart({
    required this.items,
    required this.subtotal,
    required this.hasUnavailable,
    required this.unavailable,
  });

  final List<ServerCartItem> items;
  final double subtotal;

  /// Something in the cart sold out while it sat there.
  final bool hasUnavailable;
  final List<String> unavailable;

  bool get isEmpty => items.isEmpty;
  int get sackCount => items.fold(0, (sum, i) => sum + i.quantity);

  /// Flat for now, and only charged on a cart with something in it.
  double get deliveryFee => items.isEmpty ? 0 : 80;
  double get total => subtotal + deliveryFee;

  static const empty = ServerCart(
    items: [],
    subtotal: 0,
    hasUnavailable: false,
    unavailable: [],
  );

  factory ServerCart.fromJson(Map<String, dynamic> json) {
    return ServerCart(
      items: (json['items'] as List?)
              ?.whereType<Map<String, dynamic>>()
              .map(ServerCartItem.fromJson)
              .toList() ??
          const [],
      subtotal: (json['subtotal'] as num?)?.toDouble() ?? 0,
      hasUnavailable: json['hasUnavailable'] == true,
      unavailable:
          (json['unavailable'] as List?)?.map((e) => e.toString()).toList() ??
              const [],
    );
  }
}

/// The cart, kept on the server.
///
/// It used to live only in the app's memory, which meant checkout - reading
/// the server's cart - always found it empty. It also meant a cart vanished on
/// reinstall and never followed anyone to another device.
///
/// Every call returns the whole cart, rebuilt against live products, so a
/// price that moved or an item that sold out shows up without asking.
class CartService {
  CartService._();

  static final CartService instance = CartService._();

  final _api = ApiClient.instance;

  ServerCart _parse(dynamic body) {
    if (body is! Map<String, dynamic>) return ServerCart.empty;
    return ServerCart.fromJson(body);
  }

  Future<ServerCart> fetch() async => _parse(await _api.get('/cart'));

  Future<ServerCart> add({
    required String productId,
    required double weightKg,
    required int quantity,
  }) async {
    return _parse(await _api.post('/cart/items', {
      'productId': productId,
      'weightKg': weightKg,
      'quantity': quantity,
    }));
  }

  Future<ServerCart> setQuantity({
    required String productId,
    required double weightKg,
    required int quantity,
  }) async {
    return _parse(await _api.put('/cart/items/$productId', {
      'weightKg': weightKg,
      'quantity': quantity,
    }));
  }

  Future<ServerCart> remove({
    required String productId,
    required double weightKg,
  }) async {
    return _parse(
      await _api.delete('/cart/items/$productId?weightKg=$weightKg'),
    );
  }

  Future<ServerCart> clear() async => _parse(await _api.delete('/cart'));
}

File: backend/src/services/cartService.js
import crypto from 'crypto';
import * as cartRepo from '../repositories/cartRepository.js';
import Product from '../models/Product.js';
import Order from '../models/Order.js';
import { findUserByUserId } from '../repositories/userRepository.js';

/** Two lines are the same line only if the rice and the sack size both match. */
const sameLine = (item, productId, weightKg) =>
  String(item.productId) === String(productId) &&
  Number(item.weightKg || 1) === Number(weightKg);

/** Price of one sack at this weight, applying its tier discount if one is set. */
function priceForWeight(product, weightKg) {
  const base = Number(product.price || 0) * Number(weightKg);
  const tier = (product.weightTiers || []).find((t) => Number(t.weightKg) === Number(weightKg));
  if (!tier?.discountPercent) return base;
  return Math.round(base * (1 - Number(tier.discountPercent) / 100) * 100) / 100;
}

/**
 * Rebuilds the cart against live products so the buyer sees the truth: a price
 * that moved, or an item that sold out while it sat in the cart.
 */
async function decorate(cart) {
  const ids = cart.items.map((i) => i.productId);
  const products = await Product.find({ _id: { $in: ids } }).lean();
  const byId = new Map(products.map((p) => [String(p._id), p]));

  const items = [];
  let subtotal = 0;

  for (const item of cart.items) {
    const product = byId.get(String(item.productId));
    if (!product) continue; // listing was deleted

    const weightKg = Number(item.weightKg || 1);
    const sacks = Number(item.quantity);

    // Priced per sack and multiplied, not priced as one big weight: three
    // 50 kg sacks earn the 50 kg discount three times, where 150 kg matches
    // no tier at all and would quietly lose it.
    const perSack = priceForWeight(product, weightKg);
    const lineTotal = Math.round(perSack * sacks * 100) / 100;

    const needed = weightKg * sacks;
    const available = Number(product.stock || 0);

    items.push({
      productId: String(product._id),
      sellerId: product.createdBy,
      name: product.name,
      variety: product.variety || '',
      image: product.images?.[0] || null,
      weightKg,
      quantity: sacks,
      // What the buyer's sacks come to in stock terms, so the seller's page
      // and the buyer's are talking about the same number.
      totalKg: needed,
      unitPrice: perSack,
      unitPriceAtAdd: item.unitPriceAtAdd,
      priceChanged: Number(item.unitPriceAtAdd) !== perSack,
      lineTotal,
      inStock: available >= needed,
      availableStock: available,
    });

    if (available >= needed) subtotal += lineTotal;
  }

  const blocked = items.filter((i) => !i.inStock);

  return {
    items,
    subtotal: Math.round(subtotal * 100) / 100,
    itemCount: items.reduce((n, i) => n + 1, 0),
    hasUnavailable: blocked.length > 0,
    unavailable: blocked.map((i) => i.name),
  };
}

export const getCart = async (buyerUserId) => decorate(await cartRepo.getOrCreate(buyerUserId));

export const addItem = async (buyerUserId, { productId, weightKg, quantity }) => {
  const sacks = Number(quantity);
  const weight = Number(weightKg || 1);

  if (!productId) throw new Error('Which product?');
  if (!Number.isFinite(weight) || weight <= 0) throw new Error('Choose a weight.');
  if (!Number.isFinite(sacks) || sacks <= 0) throw new Error('Choose how many.');

  const product = await Product.findById(productId);
  if (!product) throw new Error('Product not found');
  if (product.status === 'inactive') throw new Error('That product is not for sale right now.');

  const cart = await cartRepo.getOrCreate(buyerUserId);
  const existing = cart.items.find((i) => sameLine(i, productId, weight));

  // Checked against what the cart would hold in total, not just what is being
  // added, or three separate taps could add past the stock one at a time.
  const wanted = weight * (sacks + Number(existing?.quantity || 0));
  if (Number(product.stock) < wanted) {
    throw new Error(`Only ${product.stock} kg left.`);
  }

  if (existing) {
    existing.quantity = Number(existing.quantity) + sacks;
  } else {
    cart.items.push({
      productId: product._id,
      sellerId: product.createdBy,
      weightKg: weight,
      quantity: sacks,
      unitPriceAtAdd: priceForWeight(product, weight),
    });
  }

  await cartRepo.save(cart);
  return decorate(cart);
};

export const setQuantity = async (buyerUserId, productId, quantity, weightKg = 1) => {
  const sacks = Number(quantity);
  const weight = Number(weightKg || 1);

  const cart = await cartRepo.getOrCreate(buyerUserId);
  const item = cart.items.find((i) => sameLine(i, productId, weight));
  if (!item) throw new Error('That item is not in your cart.');

  if (sacks <= 0) {
    cart.items = cart.items.filter((i) => !sameLine(i, productId, weight));
  } else {
    const product = await Product.findById(productId);
    if (product && Number(product.stock) < weight * sacks) {
      throw new Error(`Only ${product.stock} kg left.`);
    }
    item.quantity = sacks;
  }

  await cartRepo.save(cart);
  return decorate(cart);
};

export const removeItem = async (buyerUserId, productId, weightKg = 1) =>
  setQuantity(buyerUserId, productId, 0, weightKg);

export const clearCart = async (buyerUserId) => decorate(await cartRepo.clear(buyerUserId));

/**
 * Turns the cart into orders. One Order row per line item — sellers fulfil and
 * get paid separately — but all rows share a groupId and order number so the
 * buyer sees one receipt.
 */
export const checkout = async (buyerUserId, {
  customerName,
  customerContact,
  deliveryAddress,
  paymentMethod,
  notes,
  deliveryFee = 0,
  paymentProof = '',
  paymentReference = '',
}) => {
  const cart = await cartRepo.getOrCreate(buyerUserId);
  if (!cart.items.length) throw new Error('Your cart is empty.');

  const view = await decorate(cart);
  if (view.hasUnavailable) {
    throw new Error(`Out of stock: ${view.unavailable.join(', ')}. Remove them to continue.`);
  }
  if (!deliveryAddress?.trim()) throw new Error('Add a delivery address.');
  if (!customerName?.trim()) throw new Error('Add the name for this delivery.');

  const buyer = await findUserByUserId(buyerUserId);
  const groupId = crypto.randomUUID();
  const orderNumber = `AGF-${crypto.randomBytes(3).toString('hex').toUpperCase()}`;

  // The delivery fee belongs to the whole basket, so it lands on the first row
  // only — otherwise the buyer is charged once per seller.
  const orders = [];
  for (const [index, line] of view.items.entries()) {
    const fee = index === 0 ? Number(deliveryFee) : 0;
    orders.push(await Order.create({
      sellerId: line.sellerId,
      productId: line.productId,
      // The name carries the sack size, since an order row for "3 × Jasmine"
      // does not say whether that is 3 kg or 150.
      productName: line.weightKg > 1
        ? `${line.name} (${line.weightKg} kg)`
        : line.name,
      unitPrice: line.unitPrice,
      quantity: line.quantity,
      subtotal: line.lineTotal,
      deliveryFee: fee,
      total: line.lineTotal + fee,
      paymentMethod: paymentMethod || 'Cash/COD',

      // A receipt attached at checkout means the buyer has already sent the
      // money and is waiting on the seller to confirm it. Without one the
      // order is simply unpaid - cash on delivery, or GCash still to be sent.
      paymentStatus: paymentProof ? 'proof_sent' : 'unpaid',
      paymentProof,
      paymentReference,
      customerName: customerName.trim(),
      customerContact: customerContact?.trim() || buyer?.contact || '',
      deliveryAddress: deliveryAddress.trim(),
      buyerUserId,
      status: 'pending',
      source: 'storefront',
      orderNumber,
      groupId,
      notes: notes?.trim() || '',
    }));
  }

  await cartRepo.clear(buyerUserId);

  return {
    orderNumber,
    groupId,
    itemCount: orders.length,
    total: orders.reduce((sum, o) => sum + Number(o.total), 0),
    paymentMethod: paymentMethod || 'Cash/COD',
    deliveryAddress: deliveryAddress.trim(),
  };
};
